### PR TITLE
Dealing with a canceled Context

### PR DESCRIPTION
**Pull request by Deleplace**
Hi Seth, I spotted a small bug with a straightforward fix so I'm taking the liberty to make a PR without creating an issue first.

## Current behaviour

- `NewConstant` rejects a duration of `0`
- It is however possible that a `Backoff.Next()` returns a duration of `0`, e.g. when using `WithJitter`
- `Do` executes the user func at least once, even if the provided Context is already canceled

## Problems

- When the backoff duration is `0`, it is possible that the user func gets retried, bypassing the Context cancelation check. This is because Go's `select` statement doesn't let us specify case priority, when several cases are ready to proceed.
- It seems to me that, when the provided Context is already canceled, `Do` should not execute the user func at all.

## Fix

- Test `ctx.Done()` alone, before the `select` with 2 cases.
- Test `ctx.Done()` before the very first call to the user func.

For readability, please find 3 commits. The 1st commit is a test that usually fails, before the code fix.

### CODE REVIEW

File: retry.go [hunk 1]
```diff
@@ -66,6 +73,13 @@ func Do(ctx context.Context, b Backoff, f RetryFunc) error {
 			return rerr.Unwrap()
 		}
 
+		// ctx.Done() has priority, so we test it alone first
```
Comment (sethvargo) [inline retry.go:76]: If only Go had a priority for `select`...

**Review by sethvargo — APPROVED** [review]
This LGTM, but can you move the test into an existing test file (maybe retry_test.go)?

**Comment by sethvargo** [comment]
> It is however possible that a Backoff.Next() returns a duration of 0, e.g. when using WithJitter

Can you tell me more about this?

**Comment by Deleplace** [comment]
WithJitter may produce a duration 0 when the magnitude of the jitter exceeds the duration of the "next backoff": [val = 0](https://github.com/sethvargo/go-retry/blob/eb868fa054ca14fa51f10f26ac7e6039e38cecb4/backoff.go#L43)

This case does not produce a hot loop though, because statistically there is more delay added by positive jitter, than subtracted by negative jitter.

**Comment by Deleplace** [comment]
Test moved to retry_test.go

**Pull request merged**